jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java

```diff
@@ -0,0 +1,266 @@
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.decomposition;
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.util.*;
+import org.jgrapht.graph.*;
+
+/**
+ * An abstract builder class for nice tree decompositions, which builds the tree decomposition in a
+ * top-down manner.
+ * <p>
+ * A tree decomposition of a graph $G$ is a tree $T$ and a map $b:V(T) \rightarrow Set&lt;V(G)&gt;$,
+ * which satisfies the properties:
+ * <ul>
+ * <li>for every edge $e \in E(G)$, there is a node $t \in V(T)$ with $e \subseteq b(v)$</li>
+ * <li>for all vertices $v \in V(G)$ the set $\{t \in V(T) | v \in b(t)\}$ is non-empty and
+ * connected in $T$</li>
+ * </ul>
+ * <br>
+ * A nice tree decomposition is a special tree decomposition, which satisfies the properties:
+ * <ul>
+ * <li>for root $r \in V(T)$ and leaf $l \in V(T): b(r)=b(t)=\emptyset$</li>
+ * <li>every non-leaf node $t \in V(T)$ is of one of the following three types:
+ * <ul>
+ * <li>forget node: $t$ has exactly one child $d$ and $b(t) \cup \{ w\} = b(d)$ for some $w \in
+ * V(G)\setminus b(t)$</li>
+ * <li>introduce node: $t$ has exactly one child $d$ and $b(t) = b(d) \cup \{ w\}$ for some $w \in
+ * V(G)\setminus b(d)$</li>
+ * <li>join node: $t$ has exactly two children $d_1$, $d_2$ and $b(t)=b(d_1)=b(d_2)$</li>
+ * </ul>
+ * </ul>
+ * <p>
+ * See:<br>
+ * Bodlaender, Hans &amp; Kloks, Ton. (1991). Better Algorithms for the Pathwidth and Treewidth of
+ * Graphs. 544-555. 10.1007/3-540-54233-7_162.
+ * 
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ *
+ * @param <V> the vertices of the graph
+ */
+abstract public class NiceDecompositionBuilder<V>
+{
+
+    // resulting decomposition
+    protected Graph<Integer, DefaultEdge> decomposition;
+
+    // map from decomposition nodes to the vertex sets
+    protected Map<Integer, Set<V>> decompositionMap;
+
+    // the root of the tree
+    protected Integer root;
+
+    // next integer for vertex generation
+    private Integer nextInteger;
+
+    /**
+     * Constructor for all methods used in the abstract method. This constructor instantiates the
+     * tree of the decomposition, the map from tree vertices to vertex sets and adds the root to the
+     * tree.
+     */
+    protected NiceDecompositionBuilder()
+    {
+        // creating objects
+        decomposition = new DefaultDirectedGraph<Integer, DefaultEdge>(DefaultEdge.class);
+        decompositionMap = new HashMap<Integer, Set<V>>();
+
+        // create root
+        root = 0;
+        nextInteger = 1;
+        decompositionMap.put(root, new HashSet<V>());
+        decomposition.addVertex(root);
+    }
+
+    /**
+     * Getter for the next free Integer. Supplies the add vertex methods with new vertices
+     * 
+     * @return unused integer
```

Comment:
Use either free or unused or next. And make sure it is clear what you mean by it.